ipaserver/install/installutils.py

```diff
@@ -436,6 +438,71 @@ def unquote_directive_value(value, quote_char):
     return unescaped_value
 
 
+SENTINEL = object()
+
+
+class DirectiveSetter(object):
+    """Safe directive setter
+    
+    with DirectiveSetter('/path/to/conf') as ds:
+        ds.set(key, value)
+    """
+    def __init__(self, filename, quotes=True, separator=' '):
+        self.filename = os.path.abspath(filename)
+        self.quotes = quotes
+        self.separator = separator
+        self.lines = None
+        self.stat = None
+
+    def __enter__(self):
+        with io.open(self.filename) as f:
+            self.stat = os.fstat(f.fileno())
+            self.lines = list(f)
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        if exc_type is not None:
+            # something went wrong, reset
+            self.lines = None
+            self.stat = None
+            return
+
+        directory, prefix = os.path.split(self.filename)
+        # use tempfile in same directory to have atomic rename
+        fd, name = tempfile.mkstemp(prefix=prefix, dir=directory, text=True)
+        with io.open(fd, mode='w', closefd=True) as f:
+            for line in self.lines:
+                if not isinstance(line, six.text_type):
+                    line = line.decode('utf-8')
+                f.write(line)
+            self.lines = None
+            os.fchmod(f.fileno(), stat.S_IMODE(self.stat.st_mode))
+            os.fchown(f.fileno(), self.stat.st_uid, self.stat.st_gid)
+            self.stat = None
+            # flush and sync tempfile inode
+            f.flush()
+            os.fsync(f.fileno())
+
+        # rename file and sync directory inode
+        os.rename(name, self.filename)
+        dirfd = os.open(directory, os.O_RDONLY | os.O_DIRECTORY)
+        try:
+            os.fsync(dirfd)
+        finally:
+            os.close(dirfd)
+
+    def set(self, directive, value, quotes=SENTINEL, separator=SENTINEL):
+        if quotes is SENTINEL:
```

Comment:
What's the advantage of a sentinel `object` over plain old `None`?

Reply:
You can pass in ``None`` to override the default. With ``None`` as default method argument, there is no way to distinguish between ``set(k, v)`` and ``set(k, v, quotes=None)``.

Reply:
But `None` is not a valid value for either `quotes` or `separator` parameters.
I'd rather just use `None` to indicate "existing or default setting" and instead
add check for `isinstance(quotes, bool)` and `isinstance(separator, str) and len(separator) > 0`.